### PR TITLE
Python: Add sequential planner to python

### LINKED ISSUE #1915: Python: Add sequential planner to python
from #1770 

image

### PR DESCRIPTION
**Pull request by joowon-dm-snu**
### Motivation and Context

fixes #1915 

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Comment by joowon-dm-snu** [comment]
linting is failed by test params below.
image

**Comment by joowon-dm-snu** [comment]
@alexchaomander  it's done but need some decisions on directory & file names.
I try to follow C# as much as possible.

**Comment by alexchaomander** [comment]
@joowon-dm-snu With #1907 merged in, we'll make sure to review this in detail this week!

**Review by awharrison-28 — CHANGES_REQUESTED** [review]
Hi @joowon-dm-snu I'm currently reviewing this PR. 

While in progress, can you revert the changes to the launch.json file? Also, if you have the time, it'd be great to add a small example file to the `python/samples/kernel-syntax-examples` folder?

**Comment by joowon-dm-snu** [comment]
> Hi @joowon-dm-snu I'm currently reviewing this PR.
> 
> While in progress, can you revert the changes to the launch.json file? Also, if you have the time, it'd be great to add a small example file to the `python/samples/kernel-syntax-examples` folder?

@awharrison-28 
Of course. Are there any other review items? I'll gather them and process them all at once.

File: python/tests/integration/planning/sequential_planner/test_sequential_planner.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            "Write a novel outline.",
+            "NovelOutline",
+            "WriterSkill",
+            "<!--===ENDPART===-->",
+        )
+    ],
+)
+@pytest.mark.asyncio
+async def test_create_plan_with_defaults_async(
+    get_aoai_config, prompt, expected_function, expected_skill, expected_default
```
Comment (awharrison-28) [inline python/tests/integration/planning/sequential_planner/test_sequential_planner.py:86]: It looks like expected_default is not being used (though commented out).

File: python/tests/integration/planning/sequential_planner/test_sequential_planner.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    plan = await planner.create_plan_async(prompt)
+
+    # Assert
+    assert any(
+        step.name == expected_function and step.skill_name == expected_skill
+        for step in plan._steps
+    )
+
+
+@pytest.mark.parametrize(
```
Comment (awharrison-28) [inline python/tests/integration/planning/sequential_planner/test_sequential_planner.py:73]: This test in particular is quite flaky. Sometimes it passes, but more often not. Trying to figure out why.

**Review by awharrison-28 — CHANGES_REQUESTED** [review]

File: python/tests/integration/planning/sequential_planner/test_sequential_planner.py (hunk 2 above)
Comment (joowon-dm-snu) [reply python/tests/integration/planning/sequential_planner/test_sequential_planner.py:73]: I've run this test many times and it always passes 😭  
Could you show me an error log for that failure?

File: python/semantic_kernel/planning/plan.py [hunk 3]
```diff
@@ -381,7 +391,7 @@ def get_next_step_variables(
         # - Step Parameters (pull from variables or state by a key value)
         function_params = step.describe()
         for param in function_params._parameters:
-            if param.name.lower == "input":
+            if param.name.lower() == "input":
```
Comment (awharrison-28) [inline python/semantic_kernel/planning/plan.py:394]: great catch here. This hid bugs in string checking, so I'm pushing some changes to resolve and fix the unit tests. 

There are several areas where I changed the pattern like below:
```
-        if input is not None:
+        if input is not None and input != "":
```

**Comment by awharrison-28** [comment]
@joowon-dm-snu I addressed the linting errors and made a few small changes to address the unit tests failing (downstream affect from the bug you fixed) - if you're good with the changes, go ahead and resolve the conversation I added.

LGTM - would love to get this in :)

**Review by awharrison-28 — APPROVED** [review]

**Comment by joowon-dm-snu** [comment]
@awharrison-28 
made 2 changes.

1) fix wrong iteration
2) sync planner to [link](https://github.com/microsoft/semantic-kernel/pull/1195/files#diff-a78ab536ff9fc4cbf9edca0434c0e61e280d0d73542dfeba38dffe4b867b5b16)

**Comment by joowon-dm-snu** [comment]
and i think planner object still need more improvements on its logic.
However, since there are no issues with the sequential planner itself, it would be beneficial to thoroughly examine it in another issue after merging this PR

**Review by awharrison-28 — APPROVED** [review]
I modified the input variable checking - works much better now and changed the syntax example prompt to something that runs more reliably - there are some issues around variable typing for mathskill that should be addressed in another PR.

**Review by awharrison-28 — APPROVED** [review]

**Pull request merged**